spring-kafka/src/test/java/org/springframework/kafka/core/KafkaTemplateTransactionTests.java

```diff
@@ -269,6 +270,37 @@ public void testTransactionSynchronizationExceptionOnCommit() {
 		assertThat(producer.closed()).isTrue();
 	}
 
+	@Test
+	public void testDeadLetterPublisherWhileTransactionActive() {
+		@SuppressWarnings("unchecked")
+		Producer<Object, Object> producer1 = mock(Producer.class);
+		@SuppressWarnings("unchecked")
+		Producer<Object, Object> producer2 = mock(Producer.class);
+		producer1.initTransactions();
+
+		@SuppressWarnings("unchecked")
+		ProducerFactory<Object, Object> pf = mock(ProducerFactory.class);
+		given(pf.transactionCapable()).willReturn(true);
+		given(pf.createProducer()).willReturn(producer1);
+
+		KafkaTemplate<Object, Object> template = new KafkaTemplate<>(pf);
+		template.setDefaultTopic(STRING_KEY_TOPIC);
+
+		KafkaTransactionManager<Object, Object> tm = new KafkaTransactionManager<>(pf);
+
+		new TransactionTemplate(tm).execute(s -> {
+			new DeadLetterPublishingRecoverer(template).accept(
+					new ConsumerRecord<>(STRING_KEY_TOPIC, 0, 0L, "key", "foo"),
+					new RuntimeException("foo"));
+			return null;
+		});
+
+		verify(producer1).beginTransaction();
+		verify(producer1).commitTransaction();
+		verify(producer1).close();
+		verify(producer2, never()).beginTransaction();
```

Comment:
The `producer2` is really never used. Not sure why we need it at all in the test.
Just because you have only this:
```
given(pf.createProducer()).willReturn(producer1);
```

On the other hand since we really talk here about `executeInTransaction()` from the `DeadLetterPublishingRecoverer`, then it would be better to verify exactly this method call. Maybe really with such a `never()`...